src/commands/snapshot/source/nft-topu.ts

```diff
@@ -0,0 +1,227 @@
+    const number = Number(match[1]);
+    return Number.isFinite(number) ? number : 0;
+  }
+
+  return 0;
+}
+
+function calculateBaseTokenBalance(counts: NftCounts): number {
+  return counts.epic * 1_000_000 + counts.normal;
+}
+
+function processNftItem(item: SearchAssetsResultItem, ownerToCounts: Map<string, NftCounts>): void {
+  const owner = item.ownership?.owner;
+  if (!owner) return;
+
+  const name = item.content?.metadata?.name;
+  const nftNumber = extractNftNumber(name);
+  const isEpic = epicNFTNumberSet.has(nftNumber);
+
+  const counts = ownerToCounts.get(owner) || { epic: 0, normal: 0 };
+  if (isEpic) {
+    counts.epic += 1;
+  } else {
+    counts.normal += 1;
+  }
+
+  ownerToCounts.set(owner, counts);
+}
+
+async function fetchPage(rpc: string, page: number): Promise<SearchAssetsResultItem[]> {
+  try {
+    const data = await retry(CONSTANTS.RETRY_CONFIG, async () => {
+      const response = await fetch(rpc, {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({
+          jsonrpc: '2.0',
+          id: `nft-topu.searchAssets.page.${page}`,
+          method: 'searchAssets',
+          params: {
+            grouping: ['collection', CONSTANTS.NFT_COLLECTION_ID],
+            page,
+            limit: CONSTANTS.PAGE_LIMIT,
+          },
+        }),
+      });
+
+      if (!response.ok) {
+        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
+      }
+
+      const body: RpcResponse = await response.json();
+      if (body.error) {
+        throw new Error(`RPC Error: ${body.error.message || 'Unknown error'}`);
+      }
+
+      return body.result || { items: [] };
+    });
+
+    return data.items || [];
+  } catch (error) {
+    throw new Error(`Failed to fetch page ${page}: ${(error as Error).message}`);
+  }
+}
+
+async function collectNftData(rpc: string): Promise<Map<string, NftCounts>> {
+  const ownerToCounts = new Map<string, NftCounts>();
+  let page = 1;
+
+  while (true) {
+    const items = await fetchPage(rpc, page);
+
+    for (const item of items) {
+      processNftItem(item, ownerToCounts);
+    }
+
+    if (items.length < CONSTANTS.PAGE_LIMIT) {
+      break;
+    }
+
+    page += 1;
+  }
+
+  return ownerToCounts;
+}
+
+function generateSnapshots(ownerToCounts: Map<string, NftCounts>, opts: StreamOptions): void {
+  let totalBaseTokenBalance = 0n;
+
+  for (const [owner, counts] of ownerToCounts.entries()) {
+    const baseTokenBalance = calculateBaseTokenBalance(counts);
+    totalBaseTokenBalance += BigInt(baseTokenBalance);
+    opts.produceSnapshot({ owner, baseTokenBalance });
+  }
+
+  if (totalBaseTokenBalance !== CONSTANTS.EXPECTED_TOTAL_BASE_TOKEN_BALANCE) {
+    throw new Error(
+      `Total amount mismatch: got ${totalBaseTokenBalance}, expected ${CONSTANTS.EXPECTED_TOTAL_BASE_TOKEN_BALANCE}`,
+    );
+  }
+}
+
+export const NftTopu: SourceStreamFactory = async (opts: StreamOptions) => {
+  try {
```

Comment:
need to validate pre-defined `virtual receipt token mint` and `pool address (collection?)` from arguments